<table>
    <tr>
  
</table>

# <FONT SIZE=5 COLOR="green"> 1. Python </FONT>

Python es un lenguaje de programación de propósito general, usado ampliamente para

*  Desarrollo de aplicaciones
*  Desarrollo web
*  Web Scraping
*  **Ciencia de Datos**
*  **Visualización**
*  Aplicaciones a los negocios
*  **Machine Learning**
*  **Deep Learning**

Python cuenta con librerías especializadas dependiendo del tipo de tarea. Algunas de ellas son:

<center><img src="https://github.com/Fabian830348/cursos/blob/master/librerias.png?raw=true" alt="centered image" width="550" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: https://www.actumlogos.com/python-para-ia.html  </FONT> <figcaption></center>

1.   **Pandas:** manipulación de datos-estadísticas-data.frames
2.   **Matplotlib**: visualización-gráficas
3.   **numpy**: operaciones matemáticas
4.   **spicy**: realizar cálculo simbólico
5.   **scikit-learn**: ejecutar algoritmos de machine learning
6.   **NTKL**: Procesamiento de Lenguaje Natural.
7.   **seaborn**: creación de gráficos estadísticos
8.   **Datetime**: para manipulación de fechas



---



<center><FONT SIZE=5 COLOR="purple"> EXPLORACIÓN DE DATOS EN PYTHON</FONT></center>

<center><img src="https://miro.medium.com/v2/resize:fit:720/format:webp/0*fto5mNElXLZ8TCZr.png" alt="centered image" width="500" height="250"></center>

+ Librería de Python especializada manipulación y análisis de estructuras de datos
+ Permite leer datos en diferentes formatos: csv, excel, bases de datos SQL.
+ Realiza operaciones rápidas y eficientes sobre conjuntos de datos
+ Permite trabajar series de tiempo
+ *import pandas as pd*
+ [Guía Pandas](https://aprendeconalf.es/docencia/python/manual/pandas/)



<center><FONT SIZE=5 COLOR="blue"> ¿Cómo importamos una librería para trabajar?</FONT></center>

In [1]:
import pandas as pd

In [2]:
import numpy as np



---



<center><FONT SIZE=5 COLOR="blue"> DATAFRAMES </FONT></center>

Los tipos de datos que vamos a trabajar se denominan data.frames

<center><img src="https://github.com/Fabian830348/cursos/blob/master/dataframe.png?raw=true" alt="centered image" width="600" height="300"></center>

Observe que un data.frame puede tener diferentes tipos de variables

<center><img src="https://github.com/Fabian830348/cursos/blob/master/tiposvariables.png?raw=true" alt="centered image" width="600" height="400"></center>





---



<center><FONT SIZE=5 COLOR="PURPLE"> PROCESO DE EXPLORACIÓN DE LOS DATOS </FONT></center>

<br>

<center><img src="https://github.com/Fabian830348/cursos/blob/master/imagen_eda.png?raw=true" alt="centered image" width="550" height="350"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia </FONT> <figcaption></center>

En el gráfico anterior se muestran algunas etapas fundamentales:

1.	**Preguntas para responder**. Es importante tener claro para qué queremos hacer el procesamiento y análisis de datos. Es decir, qué buscamos responder.

2.	**Generalidades del conjunto de datos**.  Acá buscamos reconocer la dimensión, datos faltantes, y en general, hacer un resumen del conjunto de datos.

3.	**Tipos de Variables**. Clasificar las variables en cualitativas y cuantitativas y examinar como se presenta esta información. Por ejemplo, una variable puede estar codificada y verse como numérica, pero ser categórica.

4.	**Estadística Descriptiva**. Se busca resumir la información. Utilizar medidas de tendencia, de dispersión, de forma, de posición, etc., para entender mejor los datos. Por ejemplo, se identifican datos atípicos usando boxplot.

5.	**Visualización**.  Es una etapa fundamental en el análisis de datos. Visualizar la información y poder describirla gráficamente. En esta etapa se usan: diagramas de barras, pie, boxplot, histogramas, líneas de densidad, diagramas de dispersión, entre otros.

6.	**Relación entre variable**. Se explora datos bivariados. Se examinan relaciones entre variables cualitativas (tablas de contingencia) y relaciones entre variables cuantitativas (diagramas de dispersión) se establecen posibles relaciones lineales.

7.	**Reporte de resultados**. Una vez se haga el proceso de exploración y análisis en Python, R-Studio u otro se deben reportar los resultados y conclusiones de lo que se encontró en el conjunto de datos.

Supongamos que tenemos estos datos

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/supervisado5.png?raw=true" alt="centered image" width="700" height="200"></center>
<br>

donde,

- ***Pregnancies***: Número de embarazos que ha tenido en su vida
- ***Glucose***: Nivel de concentración de glucosa en sangre
- ***BloodPressure***: Presión arterial
- ***SkinThikness***: Espesor de piel a la altura del triceps
- ***Insulin***: Respuesta a dosis de insulina en 2 horas
- ***BMI***: Índice de masa corporal
- ***DiabetesPedigreeFunction***: Presencia de diabetes en ascendencia directa
- ***Age***: Edad del paciente
- ***Outcome***: Variable que queremos predecir:
   - $1$ : Tiene diabetes
   - $0$ : No tiene diabetes

***Problema***: Predecir si una mujer, con ciertos valores en las variables antes mencionadas tiene o no diabetes.

# <FONT SIZE=5 COLOR="green"> 2. Carga de Datos en Python </FONT>

Para cargar datos en Google-Colab podemos usar varias alternativas, por ejemplo

- Desde el drive

- Repositorio externo como Github u otro.

Se pueden cargar datos en diferentes formatos, por ejemplo:

- read_csv : archivos separados por comas

- read_excel: clásicas hojas de cálculo

- read_json : javascript object notation. (formato java)

## <FONT SIZE=4 COLOR="magenta"> 2.1 **Forma 1: Desde el drive** </FONT>

Con la siguiente instrucción conectamos google-colab con Drive para poder cargar archivos desde este último.

In [ ]:
# para acceder a los datos del drive
from google.colab import drive
drive.mount('/content/drive')

Ya hemos importado la librería ***pandas*** previamente.

Vamos a importar el conjunto de datos denominado APARTAMENTOS.

$$pd.read\_csv(''ruta'')$$

In [ ]:
# cargamos los datos
aptos = pd.read_csv("/content/drive/MyDrive/Apartamentos.csv")


In [ ]:
# para ver el tipo de datos
type(aptos)

In [ ]:
# si ponemos el nombre de los datos se muestra el data.frame
aptos

**Nota importante:**
Cuando cargamos datos desde el *DRIVE* las rutas de acceso de los conjuntos de datos son los correspondientes al propietario del notebook.

En ese orden de ideas, si vamos a ejecutar las líneas de código en otra cuenta debemos cambiar las rutas de acceso o de lo contrario, producirá un error.


# <FONT SIZE=5 COLOR="green"> 3. Exploración de los datos </FONT>

In [ ]:
# no se recomienda ver estos datos.
aptos

,Unnamed: 0,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
0,1,79.0,43.16,norte,3,3,1,si,si,0.050,14.923002,no
1,2,93.0,56.92,norte,2,2,1,si,si,0.069,27.000000,si
2,3,100.0,66.40,norte,3,2,2,no,no,0.000,15.738427,no
3,4,123.0,61.85,norte,2,3,2,si,si,0.130,27.000000,no
4,5,135.0,89.80,norte,4,3,2,si,no,0.000,39.567000,si
...,...,...,...,...,...,...,...,...,...,...,...,...
689,690,498.0,180.00,laureles,5,4,3,si,si,0.570,233.784000,si
690,691,508.0,209.00,laureles,6,4,3,si,si,0.450,258.850000,si
691,692,550.0,149.00,laureles,6,3,4,no,si,0.747,326.157000,si
692,693,630.0,179.00,laureles,5,3,3,si,si,0.340,324.114000,si


1. Usamos la instrucción **head()** para explorar la cabeza de los datos

- Podemos indicar cuántos datos queremos visualizar simplemente con $head(n)$, sonde $n$ indica el número de registros que queremos ver.

In [ ]:
# para ver los primeros registros
aptos.head(56)
df

2. Usamos la instrucción **tail()** para explorar los últimos datos

In [ ]:
# para ver los últimos registros
aptos.tail(4)

3. Con la instrucción **shape** miramos el tamaño de nuestros datos

In [ ]:
# para ver el tamaño de los datos
aptos.shape

(694, 12)

es decir, el conjunto de datos tiene 694 registros o filas y 12 columnas.

4. Para explorar nombre de las variables o columnas **columns**

In [ ]:
# para ver el nombre de las columnas
aptos.columns

Index(['Unnamed: 0', 'precio', 'mt2', 'ubicacion', 'estrato', 'alcobas',
       'banos', 'balcon', 'parqueadero', 'administracion', 'avaluo',
       'terminado'],
      dtype='object')

Podemos ver que tenemos columnas como el precio, la ubicación, el avalúo, si tiene o no parqueadero, etc. Sin embargo, aparece una variable que debemos eliminar (más adelante lo haremos) que se denomina *Unnamed: 0* y que no aporta información relevante.

5. **info()** es un resumen de las variables y su clasificación. Además, podemos identificar datos faltantes (NA) en cada una de ellas.

In [ ]:
# para ver datos faltantes y tipos de datos
aptos.info()

6. Para revisar las estadística generales de las variables **describe()**

- Cuantitativas : ***datos.describe()***

- Cualitativas: ***datos.describe(include='object')***

In [ ]:
# resumen estadístico variables cuantitativas
aptos.describe()

**¿Qué puede decir de la tabla anterior?**

In [ ]:
# include = "object" resumen pequeño de variables cuali
aptos.describe(include='object')

,ubicacion,balcon,parqueadero,terminado
count,694,694,694,694
unique,7,2,2,2
top,poblado,si,si,si
freq,268,509,568,592


7. Para acceder a una o varias variable del data.frame podemos usar la instrucción

$$datos[``\text{nombre de la variable}"]$$

$$datos.variable$$

veamos algunos ejemplos


In [ ]:
aptos.columns

In [ ]:
# seleccionar la variable ubicacion


In [ ]:
# seleccionar la variable precio


0      1
1      1
2      2
3      2
4      2
      ..
689    3
690    3
691    4
692    3
693    4
Name: banos, Length: 694, dtype: int64

Podemos seleccionar varias variables de la siguiente forma

In [ ]:
aptos[ ["mt2", "alcobas"] ]

In [ ]:
# seleccionar varias variables
aptos[["balcon", "precio", "mt2", "parqueadero"]]

In [ ]:
# seleccionar la variable ubicacion
# pilas: si la variable tiene espacios no aplica
aptos.banos

8. Para hacer un conteo de las categorias de una variable

$$datos[``\text{nombre de la variable}"].value\_counts()$$

veamos algunos ejemplos

In [ ]:
# conteo de las categorias de la variable ubicación


ubicacion
poblado           268
aburra sur        169
laureles           73
occidente          69
belen guayabal     67
centro             38
norte              10
Name: count, dtype: int64

# <FONT SIZE=5 COLOR="green"> 4. Manipulación de DataFrame </FONT>

En esta sección revisaremos algunas operaciones u acciones que son importantes de conocer en un Data.Frame:

- Ordenar por una variable

- Renombrar una variable

- Eliminar una variable

- Agregar una variable

- Agrupar por categorias de una variable y aplicar funciones.

Para repasar

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

<FONT SIZE=4 COLOR="magenta"> 4.1 Ordenar según una variable </FONT>

Para ordenar por una variable usamos la instrucción

$$datos.sort\_values(variable, ascending = False) $$

se usa:

- *ascending=True* :  ordena de menor a mayor.
- *ascending=False* : ordena de mayor a menor.


In [ ]:
# ordenamos por la variable Edad de menor a mayor
aptos.sort_values("precio" , ascending =  False)

,Unnamed: 0,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
346,347,1700.0,399.00,poblado,6,3,3,no,si,0.9340,701.731,si
345,346,1490.0,445.00,poblado,6,5,4,si,si,0.7490,701.731,si
344,345,1450.0,238.19,poblado,6,4,4,si,si,0.8286,675.072,si
343,344,1450.0,318.00,poblado,6,3,3,si,si,0.7490,611.027,si
342,343,1430.0,291.00,poblado,6,3,3,si,si,0.7490,701.731,si
...,...,...,...,...,...,...,...,...,...,...,...,...
348,349,63.0,43.00,aburra sur,3,3,1,no,no,0.0500,41.468,si
11,12,53.0,37.56,occidente,3,3,1,no,si,0.0000,16.486,si
10,11,47.0,43.00,occidente,2,2,1,no,no,0.0000,16.486,no
347,348,45.0,53.00,aburra sur,3,3,1,si,no,0.0000,23.899,si


In [ ]:
# ordenamos por la variable Edad de menor a mayor
aptos.sort_values("ubicacion", ascending = True).head(10)

,Unnamed: 0,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
419,420,175.0,65.0,aburra sur,3,3,2,si,si,0.154,46.291597,si
457,458,220.0,72.0,aburra sur,4,2,2,si,si,0.211,96.390800,si
458,459,220.0,73.0,aburra sur,4,3,2,si,si,0.000,96.390800,si
459,460,220.0,89.0,aburra sur,5,3,2,no,si,0.000,143.000000,si
460,461,225.0,100.0,aburra sur,3,4,3,si,si,2.280,108.600000,si
461,462,230.0,86.0,aburra sur,3,3,2,si,si,0.215,102.000000,si
462,463,230.0,86.0,aburra sur,4,3,2,si,si,0.170,94.424234,si
463,464,230.0,89.0,aburra sur,4,3,2,si,si,0.196,95.000000,si
464,465,235.0,87.0,aburra sur,4,3,2,si,si,0.147,95.000000,si
465,466,237.0,85.0,aburra sur,4,3,2,si,si,0.195,105.786000,si


<FONT SIZE=4 COLOR="magenta"> 4.2 Renombrar una variable </FONT>

Podemos renombrar una columna de la siguiente forma usando la instrucción ***rename***

$$ datos.rename(columns=\{key : value, key : value\}, inplace=True) $$

 - Para cambiar el nombre de una variable debemos usar un diccionario donde las claves son las variables actuales y los valores son los nombres nuevos de las variables.
 - Recordemos que la instrucción **inplace** se usa para determinar si cambiamos el DataFrame original.

In [ ]:
aptos.head()

In [ ]:
aptos.rename(columns = { "parqueadero" : "parking"  } )

In [ ]:
# cambiaremos el nombre de administracion por admon
aptos.rename(columns={"administracion" : "admon"}, inplace=True)

In [ ]:
aptos.rename(columns={"parqueadero" : "parking"}, inplace=True)

<FONT SIZE=4 COLOR="magenta"> 4.3 Eliminar filas y columnas</FONT>

Para eliminar una columna o una fila usamos la instrucción .drop

$$datos.drop(``nombre", axis = 1, inplace = T \, or \, F)$$

- *axis = 1* : es por columna
- *axis = 0* : es por  fila
- *inplace = T* : sobreescribe en el *df* original.
- *inplace = F* : no sobreescribe. Mantiene el original

In [ ]:
aptos.head()

,Unnamed: 0,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
0,1,79.0,43.16,norte,3,3,1,si,si,0.050,14.923002,no
1,2,93.0,56.92,norte,2,2,1,si,si,0.069,27.000000,si
2,3,100.0,66.40,norte,3,2,2,no,no,0.000,15.738427,no
3,4,123.0,61.85,norte,2,3,2,si,si,0.130,27.000000,no
4,5,135.0,89.80,norte,4,3,2,si,no,0.000,39.567000,si


In [ ]:
aptos_copy = aptos.copy()

In [ ]:
# axis = 1
aptos_copy.drop("Unnamed: 0", axis = 1 , inplace = True)

In [ ]:
# eliminar avaluo
aptos_copy.drop("avaluo", axis = 1 , inplace = False  )

,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,terminado
0,79.0,43.16,norte,3,3,1,si,si,0.050,no
1,93.0,56.92,norte,2,2,1,si,si,0.069,si
2,100.0,66.40,norte,3,2,2,no,no,0.000,no
3,123.0,61.85,norte,2,3,2,si,si,0.130,no
4,135.0,89.80,norte,4,3,2,si,no,0.000,si
...,...,...,...,...,...,...,...,...,...,...
689,498.0,180.00,laureles,5,4,3,si,si,0.570,si
690,508.0,209.00,laureles,6,4,3,si,si,0.450,si
691,550.0,149.00,laureles,6,3,4,no,si,0.747,si
692,630.0,179.00,laureles,5,3,3,si,si,0.340,si


In [ ]:
aptos_copy.head()

,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
0,79.0,43.16,norte,3,3,1,si,si,0.050,14.923002,no
1,93.0,56.92,norte,2,2,1,si,si,0.069,27.000000,si
2,100.0,66.40,norte,3,2,2,no,no,0.000,15.738427,no
3,123.0,61.85,norte,2,3,2,si,si,0.130,27.000000,no
4,135.0,89.80,norte,4,3,2,si,no,0.000,39.567000,si


In [ ]:
# elimina el nombre de la variable que le coloquemos
aptos.drop("Unnamed: 0", axis = 1 , inplace = True)

In [ ]:
aptos.head()

In [ ]:
# Elimina las filas 0 y 1 y 4.
aptos.drop([0,1,4], axis = 0 , inplace = False)

<FONT SIZE=4 COLOR="magenta"> 4.4 Agregar una variable </FONT>

Para agregar una variable usamos la instrucción

$$df[``\text{nueva variable}"] = lista $$

En este ejercicio vamos a agregar una nueva variable que sea la división de la preico y mt2. La llamaremos "precio/mt2"

In [ ]:
aptos.head()

In [ ]:
aptos["valorM2"] = aptos.precio / aptos.mt2

In [ ]:
aptos.head()

In [ ]:
aptos["precio"].mean()

<FONT SIZE=4 COLOR="magenta"> 4.5 Agrupaciones: ***groupby()*** </FONT>

- Se usa para agrupar una variable por sus categorías y luego aplicar una función a estos grupos con respecto a otra variable

<center><img src="/conter" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>


In [ ]:
aptos.groupby("ubicacion")

Observe que no se genera nada. Esto es porque internamente lo que hace pandas es agrupar por las clases o categorias de la variable.

Podemos acceder a los grupos con la instrucción

$$datos.groupby(``Variable").get\_group(``categoría")$$

In [ ]:
aptos.groupby("ubicacion").get_group("laureles")

**¿Qué podemos hacer con esto?**

Podemos hacer operaciones con las agrupaciones, así:

colocar imagen

$$datos.groupby(``Variable")[``\text{segunda variable}"].funcion()$$

Por ejemplo, supongamos que queremos agrupar por escolaridad y calcular el promedio de las edades de cada uno de los grupos.

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/groupby2.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
aptos.head()

In [ ]:
# calcula la media del precio en cada ubicación
aptos.groupby("ubicacion")["precio"].mean()

,precio
ubicacion,
aburra sur,193.672781
belen guayabal,168.865672
centro,130.815789
laureles,318.520548
norte,131.000000
occidente,164.304348
poblado,506.017317


In [ ]:
# calcular las estadísticas del precio en cada ubicación
aptos.groupby("ubicacion")["precio"].describe()

,count,mean,std,min,25%,50%,75%,max
ubicacion,,,,,,,,
aburra sur,169.0,193.672781,64.570992,45.0,140.00,185.0,250.00,330.0
belen guayabal,67.0,168.865672,50.277983,75.0,140.00,165.0,210.00,265.0
centro,38.0,130.815789,50.874438,67.0,87.00,128.0,149.25,260.0
laureles,73.0,318.520548,159.220594,25.0,235.00,290.0,370.00,1350.0
norte,10.0,131.000000,31.770007,79.0,105.75,137.5,156.25,175.0
occidente,69.0,164.304348,59.968694,47.0,120.00,160.0,210.00,320.0
poblado,268.0,506.017317,290.683116,140.0,320.00,415.0,605.00,1700.0


In [ ]:
# si agrupamos por ubicacion, y parqueadero quiero saber el valor máximo del precio
aptos.groupby(["parqueadero" , "terminado"]).precio.max()

parqueadero  terminado
no           no            260.0
             si            320.0
si           no            900.0
             si           1700.0
Name: precio, dtype: float64

In [ ]:
aptos.groupby(["ubicacion", "parqueadero"]).precio.min()

ubicacion       parqueadero
aburra sur      no              45.0
                si             112.0
belen guayabal  no              75.0
                si              86.0
centro          no              67.0
                si              70.0
laureles        no              25.0
                si             140.0
norte           no             100.0
                si              79.0
occidente       no              47.0
                si              53.0
poblado         no             170.0
                si             140.0
Name: precio, dtype: float64

In [ ]:
aptos.precio.describe()

,precio
count,694.000000
mean,317.762739
std,247.614914
min,25.000000
25%,160.000000
50%,245.000000
75%,380.000000
max,1700.000000


# <FONT SIZE=5 COLOR="green"> 5. Algunas funciones importantes </FONT>



1. Media o promedio: **.mean()**

- Es el "equilibrio" de los datos.
- Se usa para variables cuantitativas.
- Siempre está entre el mínimo y máximo de los datos.
- Es sensible a datos atípicos. Por está razón se debe tener cuidado con su interpretación.

In [ ]:
aptos.precio.mean()

317.7627391930836

2. Mediana: **.median()**

- Es el valor que divide los datos en dos partes iguales (50%-50%)
- Aplica para variables cuantitativas
- Puede o no estar en el conjunto de datos. Depende si se tiene un número par o impar de elementos.
- No se ve afectada por la existencia de datos atípicos. Por está razón en ocaciones se usa para imputación.

In [ ]:
aptos.precio.median()

245.0

3. Moda: **.mode()**

- Es el valor que más se repite. Es decir el que tiene mayor frecuencia
- Aplica para variables cualitativas y cuantitativas
- UN conjunto puede no tener o tener múltiples modas.

In [ ]:
aptos.ubicacion.mode()

,ubicacion
0,poblado


4. Máximo y mínimo del conjunto: **.max** y **.min**

In [ ]:
aptos.precio.max()

1700.0

In [ ]:
aptos.precio.min()

25.0

5. Cuartiles: $Q_1$, $Q_2$ y $Q_3$

- Se calculan para variables cuantitativas
- $Q_1$: el valor que acumula el 25% de los datos
- $Q_2$: es el valor que acumula el 50% de los datos. En ese orden de ideas es la mediana.
- $Q_3$ es el valor que acumula el 75% de los datos.
- Para calcula en Python estos valores y en general los cuantiles usamos la librería *numpy* y la instrucción:

$$ perc\_{n} = np.percentile(variable, n) $$

Veamos algunos ejemplos con la variable *precio* de *aptos*.

In [ ]:
import numpy as np

In [ ]:
print(np.percentile(aptos.precio,25))
print(np.percentile(aptos.precio,50))
print(np.percentile(aptos.precio,75))
print(np.percentile(aptos.precio,80))

160.0
245.0
380.0
421.9999999999999


# <FONT SIZE=5 COLOR="green"> 6. Algunos filtros con DataFrame </FONT>

En esta sección revisaremos algunos filtros en un dataframe. Para esto, generaremos algunas preguntas y buscaremos responderlas con los filtros.

In [ ]:
aptos.head()

,Unnamed: 0,precio,mt2,ubicacion,estrato,alcobas,banos,balcon,parqueadero,administracion,avaluo,terminado
0,1,79.0,43.16,norte,3,3,1,si,si,0.050,14.923002,no
1,2,93.0,56.92,norte,2,2,1,si,si,0.069,27.000000,si
2,3,100.0,66.40,norte,3,2,2,no,no,0.000,15.738427,no
3,4,123.0,61.85,norte,2,3,2,si,si,0.130,27.000000,no
4,5,135.0,89.80,norte,4,3,2,si,no,0.000,39.567000,si


In [ ]:
aptos.ubicacion == "norte"

<FONT SIZE=3 COLOR="blue"> Pregunta 1: ¿Cuántos apartamentos tienen precio mayor a 100 millones? </FONT>

In [ ]:
# apartamentos de mas de 100 millones
aptos[aptos.precio > 100].head()

<FONT SIZE=3 COLOR="blue"> Pregunta 2: ¿Cuántos apartamentos del poblado tienen precio mayor a 200 millones? </FONT>

In [ ]:
len(aptos[(aptos["precio"]>200) & (aptos["ubicacion"]=="poblado")])

<FONT SIZE=3 COLOR="blue"> Pregunta 3: ¿Cuántos apartamentos del sector laureles, tienen más de 50 metros cuadrados y no cuentan con parqueadero? </FONT>

In [ ]:
len(aptos[(aptos["mt2"]>50) & (aptos["ubicacion"]=="laureles") & (aptos["parqueadero"]=="no")])

De esta manera podemos seguir haciendo filtros y respondiendo algunas preguntas. Se debe tener en cuenta:

- Operadores de comparación :

    - **>** : mayor que
    - **<** : menos que
    - **>=**  : mayor o igual
    - **<=** : menor o igual
    - **!=** (diferente)

- Operadores lógicos:

    - **&**  : y
    - **|**  : o

Se deja al lector que plantee otras preguntar y las responda usando filtros.

# <FONT SIZE=5 COLOR="green"> 7. Algunas gráficas importantes </FONT>

En esta sección revisaremos algunas gráficas clásicas.

Para esto podemos usar diferentes librerías: *mathplotlib* y *plotly*. También se pueden hacer gráficos elaborados en *seaborn* Más documentación la pueden encontrar en los siguientes links.

- https://matplotlib.org/

- https://seaborn.pydata.org/

- https://plotly.com/python/plotly-express/

**Nota:** nos centraremos más en la parte gráfica y el código que en el análisis de las gráficas.

<center><FONT SIZE=5 COLOR="Magenta"> Mathplotlib - Seaborn - Plotly.Express </FONT></center>

Existen múltiples librerías para hacer gráficas en Python. En esta parte mostraremos algunos ejemplos, sin embargo, invitamos al lector a explorar con más detalle estas librerías.

- **Mathplotlib**: es una biblioteca de Python utilizada para trazar gráficos con la ayuda de otras bibliotecas como Numpy y Pandas. Es una poderosa herramienta para visualizar datos en Python. Se utiliza para crear inferencias estáticas y trazar gráficos 2D de matrices. Fue presentado por primera vez por John D. Hunter en 2002.


- **Seaborn**: también es una biblioteca de Python que se utiliza para trazar gráficos con la ayuda de Matplotlib, Pandas y Numpy. Está construido sobre el techo de Matplotlib y se considera como un superconjunto de la biblioteca Matplotlib. Ayuda a visualizar datos univariados y bivariados.

- **Plotly.express** es otra librería para visualización con gráficos interactivos que muestran algunos detalles de las gráficas.

Vamos a cargar las librerías

In [ ]:
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sb

## <FONT SIZE=4 COLOR="green"> 7.1. TABLA DE FRECUENCIA - BARRAS Y TORTAS </FONT>

Para hacer el gráfico de barras o de tortas haremos primero una tabla de frecuencias.

- Recordemos de los gráficos de barras o de tortas se usan en general para variables categóricas o para variables cuantitativas con pocos valores, como es el caso de *número de hijos*.

- En el eje $x$ van las categorías

- En el eje $y$ la cantidad o frecuencia de cada clase o categoria

- En los diagramas de barras las barras van separadas y de colocan de diferente color.

Vamos a hacer la tabla de frecuencias absolutas y los diagramas de barras y tortas de la variable *ubicacion* del conjunto de datos aparatamentos.

In [ ]:
freq = pd.DataFrame(aptos.ubicacion.value_counts())
freq

,count
ubicacion,
poblado,268
aburra sur,169
laureles,73
occidente,69
belen guayabal,67
centro,38
norte,10


Vamos a cambiar el nombre a la columna y dejarla frecuencia absoluta (freq_abs)

In [ ]:
freq.rename(columns={'count':'freq_abs'}, inplace = True)
freq

,freq_abs
ubicacion,
poblado,268
aburra sur,169
laureles,73
occidente,69
belen guayabal,67
centro,38
norte,10


In [ ]:
px.bar(x = freq.index,
       y= freq.freq_abs)

In [ ]:
px.bar(aptos.terminado)

In [ ]:
# A partir de la tabla de frecuencias de la variable ubicación
px.pie(freq,                               # tabla de frecuencias
       values= freq.freq_abs,              # valores de las frecuencias absolutas
       names= freq.index)                  # nombres de las categorias


## <FONT SIZE=4 COLOR="green"> 7.2 HISTOGRAMAS ABSOLUTO Y ACUMULADO </FONT>

- El histograma se usa para variables de tipo cuantitativo continuo o discreto con muchas mediciones diferentes (como la edad en ciertos casos).

- En el eje $x$ se identifican las clases (intervalos) de la variable.

- En el eje $y$ se identifica la frecuencia de cada una de las clases.

- Las barras en el histograma quedan contiguas y además se utiliza un solo color en estas.

In [ ]:
# Histograma más básico
px.histogram(aptos.precio)

In [ ]:
fig = px.histogram(aptos.precio,                            # variable
                   nbins = 30,                              # número de intervalos
                   x = "precio",                            # nombre de la variable x
                   title = "Histograma con edición"         # título del gráfico
                                      )

fig.update_traces(marker_color = 'cyan',                    # color de relleno
                  marker_line_color = 'red',                # color del borde
                  marker_line_width = 1,                    # grueso de la línea
                  opacity = 1)                              # transparencia del color

fig.show()

## <FONT SIZE=4 COLOR="green"> 7.3. BOXPLOT (GRÁFICO DE CAJAS Y BIGOTES) </FONT>

- El gráfico de caja, boxplot del inglés, fue ideado por Jhon Tukey en 1977 y es una representación estadística que resalta principalmente aspectos de la distribución de los datos u observaciones por medio de las medidas de posición.

- Los diagramas de caja brindan información valiosa que no se limpia en un histograma o en un polígono de frecuencias.

A continuación a describir cada uno de los elementos que tiene un boxplot

**La caja rectangular** : el gráfico contiene una caja rectangular. Es importante indicar que esta caja no tiene coordenadas cartesianas en el eje $X$  y las alturas en el eje $y$ no representa una escala de frecuencias. Las lineas horizontales en el diagrama de cajas describen el primer y tercer cuartil. $Q_1$ y $Q_3$ respectivamente.

**Mediana** : es una línea que se muestra dentro de la caja rectangular y hace referencia al valor que divide los datos en dos partes iguales. Recuerde que la mediana es igual a $Q_2$.

**Bigotes** : son lineas que salen a los costados de la caja y se usan como referencia para ubicar las observaciones que estan por fuera del $50 \%$ central de distribucion. Los bigotes nos dan información sobre la existencia de valores atípicos (*outliers*).

In [ ]:
# boxplot clásico
px.box(aptos.mt2)

In [ ]:
px.box(aptos,                                           # nombre de los datos
       y = "precio",                                    # variable cuantitativa
       x = "ubicacion",                                 # variable cualitativa o categórica
       title = "Boxplot variable precio vs ubicación",  # título del gráfico
       color = "ubicacion"                              # color por la variable categórica
             )

## <FONT SIZE=4 COLOR="green"> 7.4 DIAGRAMA DE DISPERSIÓN </FONT>

- Los diagramas de dispersión permiten representar graficamente dos variables cuantitativas.

In [ ]:
px.scatter(aptos,
           x= "precio",
           y= "mt2")

In [ ]:
px.scatter(aptos,                                     # datos
           x= "precio",                               # variable 1
           y= "mt2",                                  # variable 2
           color = "ubicacion")

In [ ]:
px.scatter(aptos,
           x= "precio",
           y= "mt2",
           color = "ubicacion",
           size = "avaluo")

In [ ]:
pip install sweetviz

In [ ]:
import sweetviz as sw

In [ ]:
reporte=sw.analyze(aptos, target_feat="precio")

In [ ]:
from IPython.display import HTML
reporte.show_html()
HTML("SWEETVIZ_REPORT.html")

# <FONT SIZE=5 COLOR="green"> 8. Adicional: Acceder a los elementos del DataFrame </FONT>

Para acceder a los elementos, o a las filas o columnas del DataFrame podemos usar:

- ***iloc*** : para acceder por la posición
- ***loc.*** : para acceder por el nombre

Si usamos la primera instrucción

- ***df.iloc[i,j]*** : elemento de la fila i y columna j.

- ***df.iloc[i]*** : devuelve la fila i

- ***df.iloc[lista i , lista j]*** : devuelve un data.frame con las filas en lista i y columnas en columna j.

- ***df.iloc[:, j]*** : devuelve toda columna j.

Si usamos la instrucción **.loc**

- ***df.loc[nombre fila , nombre columna]***: es lo mismo anterior pero con los nombres.

- ***df[nombre columna]*** : devuelve las columnas en name columnas

- ***df.columna*** : otra forma para acceder a una columna que en el nombre no tenga espacios en blanco.

## <FONT SIZE=4 COLOR="magenta"> 8.1 Ejemplos con **.iloc** </FONT>




In [ ]:
df1


<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc1.png?raw=true" alt="centered image" width="600" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# para seleccionar un elemento de la fila i, columna j.
df1.iloc[1,3]

In [ ]:
# para seleccionar una fila
df1.iloc[3]

In [ ]:
# para seleccionar una columna
df1.iloc[ : , 1]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc2.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas 1 y 3 y columnas 0 y 1
df1.iloc[[1,3], [0,1]]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc3.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# filas de la 1 a la 3 , tener en cuenta que termina un índice antes.
df1.iloc[1:4]

<center><img src="https://github.com/Fabian830348/cursos/blob/master/intro/iloc4.png?raw=true" alt="centered image" width="450" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>

In [ ]:
# columnas 1 en adelante, ojo los dos puntos significan : end
# : solos toman todas las filas
# 1: de 1 en adelante
df1.iloc[:, 1:]

Veamos algunos ejemplos adicionales.

In [ ]:
# las columnas 2-5
df1.iloc[: , 2:6]

In [ ]:
# para seleccionar la segunda, tercera y primera fila
df1.iloc[[3,4,0]]

In [ ]:
# para seleccionar la segunda y cuarta columna
df1.iloc[: , [1,3]]

## <FONT SIZE=5 COLOR="magenta">8.2 Ejemplos con **loc** </FONT>

Veamos algunos ejemplos con **.loc**.

Recordemos que en *.iloc* se busca por posición y en *.loc* por nombre.

In [ ]:
df1.columns

In [ ]:
df1.index

In [ ]:
# selecciono la fila 0 y 1
df1.loc[[0,1]]

In [ ]:
df1.loc[: , "Estatura"]

In [ ]:
df1.loc[[1,4] , ["Nombre", "Edad", "Nivel_Educativo"]]